src/Init/Prelude.lean

```diff
@@ -3466,61 +3529,94 @@ instance {ρ : Type u} {m : Type u → Type v} : MonadWithReaderOf ρ (ReaderT 
   withReader f x := fun ctx => x (f ctx)
 
 /--
-An implementation of [`MonadState`]. In contrast to the Haskell implementation,
-we use overlapping instances to derive instances automatically from `monadLift`.
+State monads provide one mutable value of a given type.
```

Comment:
It may be worth mentioning here that the value might not be actually mutable? If the monad has ST, or everything's done nice and linearly then I agree it'll execute mutably, but it's possible to write something in terms of `MonadStateOf` which, when instantiated to `StateT`, will be immutable, I think?

Reply:
This is a bit tricky, because the word "mutable" is  a bit overloaded here:
 * There's "semantic" mutability from the perspective of programs that take the monad as an abstraction. These programs can't tell how it's implemented, just whether e.g. running `set` then `get` gives the expected result.
 * There's mutable reference cells a la `ST.Ref`, where the cell contains a pointer that can be made to point to something else.
 * There's in-place modification of uniquely-referenced data in memory. This might also occur when pointing the `ST.Ref` at what is semantically a different value.

Here, it's referring to the first meaning. I'll see if I can clarify.